### PR TITLE
vaapi: support drm devices when running in vaapi-copy mode

### PR DESCRIPTION
**Pull request by oxc**
I agree that my changes can be relicensed to LGPL 2.1 or later.

--

I was trying to use my Skylake iGPU to hardware decode a HEVC file to be displayed on a monitor connected to my external Nvidia GPU (which does not support HEVC decoding).

Using --hwdec=vaapi-copy, mpv trys to access the Nvidia card using the emulated vaapi layer, and then fails with `Decoder profile 'VAProfileHEVCMain' not available.` (because that GPU cannot decode HEVC).

So I tried to force the driver by setting LIBVA_DRIVER_NAME=i965. This worked, but the driver caused a core dump with the following message:
> `mpv: intel_driver.c:100: intel_driver_init: Assertion 'VA_CHECK_DRM_AUTH_TYPE(ctx, VA_DRM_AUTH_DRI1) || VA_CHECK_DRM_AUTH_TYPE(ctx, VA_DRM_AUTH_DRI2) || VA_CHECK_DRM_AUTH_TYPE(ctx, VA_DRM_AUTH_CUSTOM)' failed.`

Searching for this message lead me to [this mailing list entry](https://lists.freedesktop.org/archives/libva/2015-November/003687.html) which confirms that the iGPU can be accessed via drm (more specifically, the `vainfo` tool can query the driver by using `--display drm`).

So I compared the vainfo code to mpv's vaapi code and noticed that for vaapi-copy, a dummy X11 display is created. This causes the intel-driver to choke because the X server is actually running on the nvidia module.

The following patch adds support for drm "displays", similar to how vainfo does it. This allows me to successfully decode the video on the iGPU using vaapi-copy and display it via the external GPU.

**However**, since the code now prefers drm displays (because otherwise the driver would again code dump on the X11 display), I'm a bit uncertain if this breaks existing use-cases, so a review regarding that concern would be helpful.
If it turns out to be a problem, the type of display could be made configurable, for example.

### CODE REVIEW

File: video/decode/vaapi.c [hunk 1]
```diff
@@ -99,7 +99,62 @@ static const struct va_native_display disp_x11 = {
...
+struct va_native_display_drm {
+    int drm_fd;
+};
+
+static void drm_destroy(struct priv *p)
+{
+    struct va_native_display_drm *native_display = p->native_display;
+    if (native_display) {
+        if (native_display->drm_fd < 0)
+            close(native_display->drm_fd);
```
Comment (wm4) [inline video/decode/vaapi.c]: Inverted condition.

File: video/decode/vaapi.c [hunk 2]
```diff
@@ -99,7 +99,62 @@ static const struct va_native_display disp_x11 = {
...
+        talloc_free(native_display);
+        p->native_display = NULL;
+    }
+}
+
+static void drm_create(struct priv *p)
+{
+    int i, drm_fd;
```
Comment (wm4) [inline video/decode/vaapi.c]: Declare variables where you use them.

Comment (oxc) [inline video/decode/vaapi.c]: Fixed.

File: video/decode/vaapi.c (hunk 1 above)
Comment (oxc) [inline video/decode/vaapi.c]: Thanks. Fixed.

**Comment by wm4** [comment]
Merged,

**Comment by fhlfibh** [comment]
@oxc 
> This allows me to successfully decode the video on the iGPU using vaapi-copy and display it via the external GPU.

How you do that exactly, i.e. what commands, options, parameters must be set for mpv with nvidia+skylake?

**Comment by oxc** [comment]
@fhlfibh I just use `--hwdec=vaapi-copy`, in my case that's all it takes.

**Pull request closed without merging**